### ISSUE TITLE
Use of oneOf to specify multiple item types in an array used as an in query parameter results in KeyError

### ISSUE THREAD

**Issue by LutzVenhofen** [spec-first/connexion#1300]
### Description
As per the OpenAPI definition [here](https://swagger.io/docs/specification/data-models/data-types/#array) we can define a mixed-type array. Connexion cannot handle mixed-types arrays when used as in query parameters.

### Expected behaviour
Connexion handles mixed type arrays when given as parameters and parses the arrays through to the endpoints, transforming the given values to the appropriate python types.

### Actual behaviour
When given an array Connexion validates the items of the array to the item type specified in the OpenAPI definition by accessing a dict param_schema [see decorators/validation.py](https://github.com/zalando/connexion/blob/master/connexion/decorators/validation.py#L61-L71) 

```python
 if param_type == "array":
       converted_params = []
        if parameter_type == "header":
            value = value.split(',')
        for v in value:
            try:
                converted = make_type(v, param_schema["items"]["type"])
            except (ValueError, TypeError):
                converted = v
            converted_params.append(converted)
        return converted_params
```

when multiple item types are defined in the API specification this will lead to a key-error as param_schema will have the structure
`{"items": {"oneOf": [{"type":"string", {"type":"boolean"}, ..]}}.` where as in the current implementation only one type is expected which leads to the following structure of param_schema `{"items": {"type": "string"}}.`. 

```
Testing started at 15:22 ...
C:\Users\lvenhofe\AppData\Local\Continuum\anaconda3\envs\template-webservice\python.exe "C:\Program Files\JetBrains\PyCharm 2019.2.4\helpers\pycharm\_jb_pytest_runner.py" --target test_endpoints.py::test_get_some_animal_details
Launching pytest with arguments test_endpoints.py::test_get_some_animal_details in C:\Users\lvenhofe\PycharmProjects\template-webservice_oneOf_array\connexion_api\tests

============================= test session starts =============================
platform win32 -- Python 3.7.4, pytest-6.0.2, py-1.9.0, pluggy-0.13.1 -- C:\Users\lvenhofe\AppData\Local\Continuum\anaconda3\envs\template-webservice\python.exe
cachedir: .pytest_cache
rootdir: C:\Users\lvenhofe\PycharmProjects\template-webservice_oneOf_array
collecting ... collected 1 item

test_endpoints.py::test_get_some_animal_details[chicken-details0] FAILED [100%]My url: get-some-details?name=chicken&details=color,size

connexion_api\tests\test_endpoints.py:65 (test_get_some_animal_details[chicken-details0])
api_client = <FlaskClient <Flask 'connexion_api.api.api'>>
animal_name = 'chicken', details = ['color', 'size']

    @pytest.mark.parametrize(
        "animal_name, details",
        [
            ("chicken", ["color", "size"])
        ]
    )
    def test_get_some_animal_details(api_client, animal_name, details):
        url = 'get-some-details?name={}&details='.format(animal_name)
        for d in details:
            url += "{},".format(d)
        url = url[:-1]
        print("My url: {}".format(url))
>       response = api_client.get(url)

test_endpoints.py:78: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
..\..\..\..\AppData\Local\Continuum\anaconda3\envs\template-webservice\lib\site-packages\werkzeug\test.py:1006: in get
    return self.open(*args, **kw)
..\..\..\..\AppData\Local\Continuum\anaconda3\envs\template-webservice\lib\site-packages\flask\testing.py:227: in open
    follow_redirects=follow_redirects,
..\..\..\..\AppData\Local\Continuum\anaconda3\envs\template-webservice\lib\site-packages\werkzeug\test.py:970: in open
    response = self.run_wsgi_app(environ.copy(), buffered=buffered)
..\..\..\..\AppData\Local\Continuum\anaconda3\envs\template-webservice\lib\site-packages\werkzeug\test.py:861: in run_wsgi_app
    rv = run_wsgi_app(self.application, environ, buffered=buffered)
..\..\..\..\AppData\Local\Continuum\anaconda3\envs\template-webservice\lib\site-packages\werkzeug\test.py:1096: in run_wsgi_app
    app_rv = app(environ, start_response)
..\..\..\..\AppData\Local\Continuum\anaconda3\envs\template-webservice\lib\site-packages\flask\app.py:2464: in __call__
    return self.wsgi_app(environ, start_response)
..\..\..\..\AppData\Local\Continuum\anaconda3\envs\template-webservice\lib\site-packages\flask\app.py:2450: in wsgi_app
    response = self.handle_exception(e)
..\..\..\..\AppData\Local\Continuum\anaconda3\envs\template-webservice\lib\site-packages\flask\app.py:1867: in handle_exception
    reraise(exc_type, exc_value, tb)
..\..\..\..\AppData\Local\Continuum\anaconda3\envs\template-webservice\lib\site-packages\flask\_compat.py:39: in reraise
    raise value
..\..\..\..\AppData\Local\Continuum\anaconda3\envs\template-webservice\lib\site-packages\flask\app.py:2447: in wsgi_app
    response = self.full_dispatch_request()
..\..\..\..\AppData\Local\Continuum\anaconda3\envs\template-webservice\lib\site-packages\flask\app.py:1952: in full_dispatch_request
    rv = self.handle_user_exception(e)
..\..\..\..\AppData\Local\Continuum\anaconda3\envs\template-webservice\lib\site-packages\flask\app.py:1821: in handle_user_exception
    reraise(exc_type, exc_value, tb)
..\..\..\..\AppData\Local\Continuum\anaconda3\envs\template-webservice\lib\site-packages\flask\_compat.py:39: in reraise
    raise value
..\..\..\..\AppData\Local\Continuum\anaconda3\envs\template-webservice\lib\site-packages\flask\app.py:1950: in full_dispatch_request
    rv = self.dispatch_request()
..\..\..\..\AppData\Local\Continuum\anaconda3\envs\template-webservice\lib\site-packages\flask\app.py:1936: in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
..\..\..\..\AppData\Local\Continuum\anaconda3\envs\template-webservice\lib\site-packages\connexion\decorators\decorator.py:48: in wrapper
    response = function(request)
..\..\..\..\AppData\Local\Continuum\anaconda3\envs\template-webservice\lib\site-packages\connexion\decorators\uri_parsing.py:144: in wrapper
    response = function(request)
..\..\..\..\AppData\Local\Continuum\anaconda3\envs\template-webservice\lib\site-packages\connexion\decorators\validation.py:360: in wrapper
    error = self.validate_query_parameter(param, request)
..\..\..\..\AppData\Local\Continuum\anaconda3\envs\template-webservice\lib\site-packages\connexion\decorators\validation.py:320: in validate_query_parameter
    return self.validate_parameter('query', val, param)
..\..\..\..\AppData\Local\Continuum\anaconda3\envs\template-webservice\lib\site-packages\connexion\decorators\validation.py:257: in validate_parameter
    converted_value = coerce_type(param, value, parameter_type, param_name)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

param = {'description': 'array of details requested', 'explode': False, 'in': 'query', 'name': 'details', ...}
value = ['color', 'size'], parameter_type = 'query', parameter_name = 'details'

    def coerce_type(param, value, parameter_type, parameter_name=None):
    
        def make_type(value, type_literal):
            type_func = TYPE_MAP.get(type_literal)
            return type_func(value)
    
        param_schema = param.get("schema", param)
        if is_nullable(param_schema) and is_null(value):
            return None
    
        param_type = param_schema.get('type')
        parameter_name = parameter_name if parameter_name else param.get('name')
        if param_type == "array":
            converted_params = []
            for v in value:
                try:
>                   converted = make_type(v, param_schema["items"]["type"])
E                   KeyError: 'type'

..\..\..\..\AppData\Local\Continuum\anaconda3\envs\template-webservice\lib\site-packages\connexion\decorators\validation.py:65: KeyError
```

A quick and dirty fix that worked for me looks like this:
```python
    if param_type == "array":
        converted_params = []
        for v in value:
            try:
                converted = make_type(v, param_schema["items"]["type"])
            except KeyError:  # we have "oneOf" -> [{"type": "string"}, {"type": "boolean"}, ...]
                for one_of_schema in param_schema["items"]["oneOf"]:
                    try:
                        converted = make_type(v, one_of_schema["type"])
                        break
                    except (ValueError, TypeError):
                        continue
                converted = v
            except (ValueError, TypeError):
                converted = v
            converted_params.append(converted)
        return converted_params
```
It now catches the KeyError, and tries out every type given for the array. If it finds a successful type we stop looking through all types. If all given types result in a ValueError or TypeError we just add v directly, which should mirror the original behaviour.

Fixing this part will then lead to another very similar KeyError thrown by connexion in [operations/openapi.py](https://github.com/zalando/connexion/blob/master/connexion/operations/openapi.py#L371-L380)

```python
def _get_val_from_param(self, value, query_defn):
        query_schema = query_defn["schema"]

        if is_nullable(query_schema) and is_null(value):
            return None

        if query_schema["type"] == "array":
            return [make_type(part, query_schema["items"]["type"]) for part in value]
        else:
            return make_type(value, query_schema["type"])
```

As we again try to access a "type" key which is not there as it is replaced by a "oneOf" key returning a list of dicts containing the types. Again I solved it for me with a quick and dirty fix:

```python
    def _get_val_from_param(self, value, query_defn):
        query_schema = query_defn["schema"]

        if is_nullable(query_schema) and is_null(value):
            return None

        if query_schema["type"] == "array":
            if "type" in query_schema["items"]:
                return [make_type(part, query_schema["items"]["type"]) for part in value]
            else:
                converted = []
                for part in value:
                    could_be_string_type = False
                    part_added = False
                    for one_of_schema in query_schema["items"]["oneOf"]:
                        if one_of_schema["type"] != "string":
                            try:
                                part_made_type = make_type(part, one_of_schema["type"])
                                converted.append(part_made_type)
                                part_added = True
                                break  # succesfully found a non string type
                            except (ValueError, TypeError):
                                continue # could be one of the other types
                        elif one_of_schema["type"] == "string":
                            could_be_string_type = True

                    if could_be_string_type and not part_added:
                        part_made_type = make_type(part, "string")
                        converted.append(part_made_type)
                return converted
        else:
            return make_type(value, query_schema["type"])
```
We try to convert all elements given in the arrays to their corresponding python types, we again try out every type specified for the multi-type array. Only string types will be done last, as we can transform most things to a string (integers, floats, booleans). My fix currently assumes that if we allow for example integers and strings in the array, that no string which could be transformed to an integer in python is given in the array. So for example `["2", 3]` will with this fix be transformed to `[2, 3]` for the python function. This may not be the desired outcome!
If a seconed issue should be opened for the follow up KeyError let me know and I will open it.

### Steps to reproduce
Any OpenAPI definition with a mixed-type array as a parameter will do.
```yaml
paths:
/get-some-details:
    get:
      operationId: connexion_api.api.endpoints.get_some_animal_details
      tags:
        - experimental
      summary: try out giving in_query array parameter
      parameters:
        - $ref: '#/components/parameters/details'
        - $ref: '#/components/parameters/name'
      responses:
        200:
          description: Gives all animal details
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/AnimalDetails'
components:
  parameters:
    details:
      name: details
      required: true
      in: query
      description: array of details requested
      schema:
        type: array
        items:
          oneOf:
            - type: boolean
            - type: string
      style: form
      explode: false
    name:
      name: name
      required: true
      in: query
      schema:
        type: string
  schemas:
    AnimalDetails:
      type: object
      required:
        - name
      properties:
        name:
          type: string
          example: chicken
        legs:
          type: integer
        hair:
          type: boolean
        feathers:
          type: boolean
        eggs:
          type: boolean
        milk:
          type: boolean
        airborne:
          type: boolean
        aquatic:
          type: boolean
```

## Other issues dealing with "oneOf"
A very similar issue seems to be #691 the mentioned fix moving the troubling definition into the paths segment did not work for the array.
Issue #1013 mentions anyOf and oneOf in the parameters section, the posted error seems not to be hinting at a KeyError
Output of the commands:

- `python --version` Python 3.7.4
- `pip show connexion | grep "^Version\:"` 
Name: connexion
Version: 2.7.0
Summary: Connexion - API first applications with OpenAPI/Swagger and Flask
Home-page: https://github.com/zalando/connexion
Author: Zalando SE
Author-email: None
License: Apache License Version 2.0
Location: c:\users\lvenhofe\appdata\local\continuum\anaconda3\envs\template-webservice\lib\site-packages
Requires: flask, requests, jsonschema, clickclick, PyYAML, inflection, openapi-spec-validator

**Comment by RobbeSneyders** [comment]
Thanks for the report and detailed info @LutzVenhofen. This is currently indeed not supported in Connexion, but any work on it would be appreciated.

**Comment by RobbeSneyders** [comment]
Closing in favor of #1569.
